tm2/pkg/bft/privval/signer/remote/client/conn.go

```diff
@@ -0,0 +1,143 @@
+package client
+
+import (
+	"net"
+	"time"
+
+	"github.com/gnolang/gno/tm2/pkg/amino"
+	r "github.com/gnolang/gno/tm2/pkg/bft/privval/signer/remote"
+)
+
+// ensureConnection tries to establish a connection with the server.
+func (rsc *RemoteSignerClient) ensureConnection() error {
+	// If the connection is already established, return.
+	if rsc.isConnected() {
+		rsc.logger.Debug("Already connected to server")
+		return nil
+	}
+
+	// Try to establish a connection with the server, retrying if necessary.
+	for try := 0; ; try++ {
+		// Ensure the client is not closed.
+		if rsc.isClosed() {
+			return nil
+		}
+
+		// Dial the server.
+		conn, err := net.DialTimeout(rsc.protocol, rsc.address, rsc.dialTimeout)
+		if err != nil {
+			rsc.logger.Warn("Failed to dial",
+				"protocol", rsc.protocol,
+				"address", rsc.address,
+				"error", err,
+			)
+
+			// If the maximum retries are not exceeded, log attempts count then retry.
+			if rsc.dialMaxRetries > 0 && try < rsc.dialMaxRetries {
+				rsc.logger.Info("Retrying to connect", "try", try+1, "maxRetry", rsc.dialMaxRetries)
+			} else if rsc.dialMaxRetries < 0 { // Retry indefinitely.
+				rsc.logger.Info("Retrying to connect", "try", try+1, "maxRetry", "unlimited")
+			} else { // Max retries exceeded.
+				return ErrMaxRetriesExceeded
+			}
+
+			// Wait for the retry interval before trying again.
+			time.Sleep(rsc.dialRetryInterval)
```

Comment:
Can't we use a time interval channel?

Reply:
Do you mean a ticker? If yes, we could, but what would be the benefit to do that?
There is no other channel to read from concurrently or context completion to wait for. 
But